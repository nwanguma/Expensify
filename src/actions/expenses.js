import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  //this dispatch argument is provided by thunk
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
});


// REMOVE_EXPENSE_ASYNC
export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }));
    })
  }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// EDIT_EXPENSE_ASYNC
export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    })
  }
}

// SET_EXPENSE
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

// SET_EXPENSES_ASYNC
export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];

      snapshot.forEach(childSnapshot => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setExpenses(expenses));
    });
  };
};
