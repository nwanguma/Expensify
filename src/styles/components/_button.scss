.btn {
  font-family: 'Lato', Arial, Helvetica, sans-serif;
  background-color: rgb(36, 36, 136);
  color: rgb(202, 196, 196);
  padding: 1rem 3rem;
  border: none;  
}

.btn--login {
  margin-top: 1.5rem;
}

.btn--logout {
  border: 1px solid rgb(202, 196, 196);
  padding: 0.5rem 1.5rem;
}

.btn--remove {
  display: inline-block;
  padding: .5rem .8rem;
  border-radius: 20px;
  text-align: center;
  width: 3rem;
  height: 3rem;
  background-color: rgba(255, 0, 0, 0.705);
  cursor: pointer;
}