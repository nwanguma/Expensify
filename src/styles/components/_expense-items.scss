.expense-item-container {
  padding: 2rem 4rem;
  width: 85%;
  margin: 0 auto;
  display: flex;
  align-items: center;
  justify-content: space-around;

  &:nth-child(even) {
    background-color: rgb(252, 251, 251);
    }

  &:nth-child(odd) {
    background-color: lavender;    
  }

  @media screen and (max-width: 900px) {
    justify-content: space-between;
    widtH: 100%;
    padding: 1rem .5rem;
  }
}


.expense__content {
  &--link {
    text-decoration: none;
    width: 40%;
    flex: 0 1 40%;

    @media screen and (max-width: 90rem) {
      font-size: 1.5rem;
    }
  }

  &--description {
    color: black;
    display: inline-block;
    padding: 0 2rem;
    font-size: 1.5rem;

    &::first-letter {
      text-transform: uppercase;
    }
  }

  &--amount {
    font-weight: bold;
    color: red;
    flex: 0 0 20%; 
    text-align: left;
    display: inline-block;
    justify-self: start;
    font-size: 1.5rem;
  }

  &--date {
    flex: 0 0 20%;
    font-size: 1.2rem;

    @media screen and (max-width: 450px) {
      display: none;      
      font-size: 1.3rem;
    }
  }
}