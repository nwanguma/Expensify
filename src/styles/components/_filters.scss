.filters-container {
  background-color: rgb(36, 36, 136);
  padding: 1rem 2rem;
}

.input-container {
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  width: 70%;
  margin: 0 auto;

  @media screen and (max-width: 850px) {
    flex-direction: column;
    width: 100%;
    align-items: center;
  }
}

.select, .search {
  height: 3rem;

  @media screen and (max-width: 850px) {
    margin-bottom: 1rem;
  }
}

