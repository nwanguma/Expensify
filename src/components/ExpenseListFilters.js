import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import uuid from 'uuid';
import 'react-dates/initialize';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
    startDateId: uuid(),
    endDateId: uuid()
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  }
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };
  render() {
    return (
      <div className="filters-container">
        <div className="input-container">
          <input
            placeholder="Search expenses"
            type="text"
            value={this.props.filters.text}
            onChange={this.onTextChange}
            className="search"
          />
          <select
            value={this.props.filters.sortBy}
            onChange={this.onSortChange}
            className="select"
          >
            <option value="date">Date</option>
            <option value="amount">Amount</option>
          </select>

          <DateRangePicker
            startDate={this.props.filters.startDate}
            startDateId={this.state.startDateId}
            endDate={this.props.filters.endDate}
            endDateId={this.state.endDateId}
            onDatesChange={this.onDatesChange}
            focusedInput={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            showClearDates={true}
            numberOfMonths={1}
            isOutsideRange={() => false}
            small={true}
          />
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
