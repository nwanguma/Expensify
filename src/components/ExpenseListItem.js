import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import { connect } from 'react-redux';
import { startRemoveExpense } from '../actions/expenses';

class ExpenseListItem extends React.Component {
  constructor(props) {
    super(props);
  }
  onRemove = () => {
    this.props.startRemoveExpense({ id: this.props.id });
  }
  render() {
    return (
      <div className="expense-item-container">
        <Link to={`/edit/${this.props.id}`} className="expense__content 
        expense__content--link" >
          <h3 className="expense__text expense__content--description">
            {this.props.description}</h3>
        </Link>
        <span className="expense__text expense__content--amount">
          {`â‚¦${numeral(this.props.amount / 100).format('0,0.00')}`}
        </span>
        <span className="expense__text expense__content--date">
          {moment(this.props.createdAt).format('MMMM Do, YYYY')}
        </span>
        <a onClick={this.onRemove} className="btn btn--remove" role="button">x</a>
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(undefined, mapDispatchToProps)(ExpenseListItem);
