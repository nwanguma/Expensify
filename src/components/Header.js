import React from 'react';
import { NavLink } from 'react-router-dom';
import { startLogout } from '../actions/auth';
import { connect } from 'react-redux';

export const Header = ({ startLogout }) => (
  <header className="header">
    <h1 className="header__heading">Expensify</h1>
    <nav className="header__nav">
      <NavLink to="/dashboard" activeClassName="is-active" className="header__nav__item header__nav__item--dashboard">Dashboard</NavLink>
      <NavLink to="/create" activeClassName="is-active" className="header__nav__item header__nav__item--create">New Expense</NavLink>
      <button onClick={startLogout} className="btn btn--logout">Log out</button>
    </nav>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => { dispatch(startLogout()) }
})

export default connect(undefined, mapDispatchToProps)(Header);
