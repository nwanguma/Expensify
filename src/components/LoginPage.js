import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => {
  return (
    <div className="login-container">
      <div className="login__btn-box">
        <h1 className="login__heading--primary">Expensify</h1>
        <p className="login__heading--secondary">Tracking your expenses like it's 1999.</p>
        <button onClick={startLogin} className="btn btn--login">Sign in with Google</button>
      </div>
    </div>
  )
};

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
})

export default connect(undefined, mapDispatchToProps)(LoginPage);