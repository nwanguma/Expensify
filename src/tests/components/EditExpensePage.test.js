import React from 'react';
import { EditExpensePage } from '../../components/EditExpensePage';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expense';

let id, startEditExpense, startRemoveExpense, history, wrapper;

beforeEach(() => {
  id = expenses[2].id;
  startEditExpense = jest.fn();
  startRemoveExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<EditExpensePage expense={expenses[2]}
    startEditExpense={startEditExpense}
    startRemoveExpense={startRemoveExpense} history={history} />);
})

test('Should render the EditExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('Should call onSubmit on the EditExpensePage', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
  expect(startEditExpense).toHaveBeenCalledWith(id, expenses[2]);
  expect(history.push).toHaveBeenCalledWith('/');
});

test('Should call onRemove on the EditExpensePage', () => {
  wrapper.find('button').prop('onClick')();
  expect(startRemoveExpense).toHaveBeenCalledWith({ id });
  expect(history.push).toHaveBeenCalledWith('/');
});