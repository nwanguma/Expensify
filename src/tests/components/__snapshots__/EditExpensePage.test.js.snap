// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 84800,
        "createdAt": "1969-12-01T00:00:00.000Z",
        "description": "Vacation",
        "id": "2",
        "note": "A trip to South Sudan",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
